walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -32,28 +38,63 @@
 
   /** Have end time for meetings be exclusive. */
   private static final boolean TIME_EXCLUSIVE = false;
+  /** Have end time for meetings be inclusive. */
+  private static final boolean TIME_INCLUSIVE = true;
 
   /** Construct possible meeting times based on the given events and attendees. */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    // Times which fultill requirements (all required attendees)
     List<TimeRange> possibleTimes = new ArrayList<>();
     // Starting out, any time is possible
     possibleTimes.add(TimeRange.WHOLE_DAY);
 
     Collection<String> requiredAttendees = request.getAttendees();
     Collection<String> optionalAttendees = request.getOptionalAttendees();
 
+    // Times which work for optional attendees, mapped by name of
+    // optional attendee
+    Map<String, List<TimeRange>> optionalTimes = new HashMap<>();
+    optionalAttendees.forEach(attendee -> {
+      // Each attendee can attend the whole day to start
+      List<TimeRange> times = new ArrayList<>();
+      times.add(TimeRange.WHOLE_DAY);
+      optionalTimes.put(attendee, times);
+    });
+
     // Look for collisions with other meetings
     events.stream().forEach(event -> {
-      if (isAttendeeOverlap(requiredAttendees, optionalAttendees, event) == Overlap.REQUIRED) {
+      Overlap eventOverlap = isAttendeeOverlap(requiredAttendees, optionalAttendees, event);
+      if (eventOverlap == Overlap.REQUIRED) {
         // We can't use this meeting time since at least a required attendee is at another meeting
         subtractTime(possibleTimes, event.getWhen());
       }
+
+      // Handle times for optional attendees
+      getOptional(optionalAttendees, event).forEach(attendee -> {
+        subtractTime(optionalTimes.get(attendee), event.getWhen());
+      });
     });
 
-    // Remove any times that are too small and then sort in ascending order
-    possibleTimes.removeIf(time -> time.duration() < request.getDuration());
-    Collections.sort(possibleTimes, TimeRange.ORDER_BY_START);
-    return possibleTimes;
+    cleanTimes(possibleTimes, request);
+    optionalTimes.values().forEach(times -> cleanTimes(times, request));
+
+    // Times that have been reconciled between optional and required attendees
+    List<TimeRange> combinedTimes =
+        new ArrayList<>(optimizeTimes(possibleTimes, optionalTimes, request));
+    cleanTimes(combinedTimes, request);
+
+    return combinedTimes;
+  }
+
+  /** Returns the optional attendees at the given event */
+  private static Collection<String> getOptional(Collection<String> optionalAttendees, Event event) {
```

Comment:
since Optional is an often used class name in Java (Optional\<T\>) could you consider using a more descriptive name for this method?